File: cpp_stuff/groups.cpp
//    Copyright 2020 sorrge
//
//    This file is part of tg_news_cluster.
//
//    tg_news_cluster is free software: you can redistribute it and/or modify
//    it under the terms of the GNU General Public License as published by
//    the Free Software Foundation, either version 3 of the License, or
//    (at your option) any later version.
//
//    tg_news_cluster is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU General Public License for more details.
//
//    You should have received a copy of the GNU General Public License
//    along with tg_news_cluster.  If not, see <https://www.gnu.org/licenses/>.
#include <tuple>
#include <iostream>
#include <vector>
#include <sstream>
#include <fstream>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <memory>
#include <shared_mutex>

#include "common.h"
#include "groups.h"

using namespace std;


unordered_map<string, pair<unordered_map<hash_t, float>, float>*> loaded_idfs;


pair<unordered_map<hash_t, float>, float>* read_idf_file(const string& file_name) {
	//cout << "Loading IDF " << file_name << endl;
	ifstream file(file_name, ios::binary);
	int num_texts;
	file.read((char*)&num_texts, sizeof(num_texts));
	auto loaded_data = new pair<unordered_map<hash_t, float>, float>;
	float max_idf = logf(num_texts);
	loaded_data->second = max_idf;

	while(true) {
		int count;
		file.read((char*)&count, sizeof(count));
		if(!file)
			break;

		int cc;
		file.read((char*)&cc, sizeof(cc));
		float cur_idf = max_idf - logf(count);
		for(int i = 0; i < cc; ++i) {
			hash_t h;
			file.read((char*)&h, sizeof(h));
			if(!file)
				break;

			loaded_data->first[h] = cur_idf;
		}
	}

	//cout << "IDF " << file_name << " loaded" << endl;
	return loaded_data;
}


void load_idf(const string& file_name) {
	if(loaded_idfs.count(file_name) > 0)
		return;

	loaded_idfs[file_name] = read_idf_file(file_name);
}


unordered_set<hash_t> get_words(const vector<string>& text) {
	unordered_set<hash_t> words;
	hash_t prev_hash = string_hash("");
	for(auto& word : text) {
		hash_t current_hash = string_hash(word);
		words.insert(current_hash);
		words.insert(hash_uint32(current_hash + prev_hash));
		prev_hash = current_hash;
	}

	return words;
}


void chunks_idf(const vector<vector<string>>& texts, const string& out_file) {
	unordered_map<hash_t, int> base_counts;
	for(auto& s : texts)
		for(auto& chunk : get_words(s))
			++base_counts[chunk];

	ofstream file(out_file, ios::binary);
	int num_texts = (int)texts.size();
	file.write((char*)&num_texts, sizeof(num_texts));
	vector<pair<hash_t, int>> base_counts_sorted(base_counts.begin(), base_counts.end());
	sort(base_counts_sorted.begin(), base_counts_sorted.end(), [](pair<hash_t, int>& p1, pair<hash_t, int>& p2)
			{ return p1.second < p2.second || (p1.second == p2.second && p1.first < p2.first); });

	unordered_map<int, int> count_of_counts;
	for(auto& kvp : base_counts_sorted)
		++count_of_counts[kvp.second];


	int prev_count = -1;
	for(auto& kvp : base_counts_sorted) {
		if(kvp.second == 1)
			continue;

		if(kvp.second != prev_count) {
			prev_count = kvp.second;
			file.write((char*)&prev_count, sizeof(prev_count));
			int cc = count_of_counts[kvp.second];
			file.write((char*)&cc, sizeof(cc));
		}

		file.write((char*)&kvp.first, sizeof(kvp.first));
	}
}


void get_words_with_count_parallel(const vector<tuple<vector<string>, string, string, double>>& article_data_text, vector<article_data>& article_data_processed) {
	article_data_processed.resize(article_data_text.size());
#pragma omp parallel for
	for(int i = 0; i < (int)article_data_text.size(); ++i) {
		unordered_map<hash_t, int> chunks;
		get_words_with_count(get<0>(article_data_text[i]), chunks);
		article_data_processed[i].chunks.reserve(chunks.size());
		for(auto& kvp : chunks)
			article_data_processed[i].chunks.push_back(kvp);

		article_data_processed[i].site = get<1>(article_data_text[i]);
		article_data_processed[i].file_name = get<2>(article_data_text[i]);
		article_data_processed[i].timestamp = get<3>(article_data_text[i]);
	}
}


vector<float> calc_popularities(const vector<tuple<vector<string>, string, string, double>>& text_data, const string& idf) {
	vector<article_data> articles;
	get_words_with_count_parallel(text_data, articles);
	const auto& idf_data = *loaded_idfs.at(idf);
	grouper g(idf_data.first, idf_data.second, 1.0f);
	g.add_texts(articles);
	vector<float> popularities(text_data.size(), 0);
//	for(auto& sim_kvp : g.get_sims()) {
//		popularities[sim_kvp.first.first] += sim_kvp.second;
//		popularities[sim_kvp.first.second] += sim_kvp.second;
//	}

	return popularities;
}


void get_words_with_count(const vector<string>& text, unordered_map<hash_t, int>& words) {
	hash_t prev_hash = 0;
	for(auto& word : text) {
		hash_t current_hash = string_hash(word);
		++words[current_hash];
		if(prev_hash != 0)
			++words[hash_uint32(current_hash + prev_hash)];

		prev_hash = current_hash;
	}
}


vector<vector<int>> group_texts(const vector<tuple<vector<string>, string, string, double>>& article_data_text, const string& idf_file, float similarity_cutoff) {
	auto idf_data = loaded_idfs.at(idf_file);
	grouper g(idf_data->first, idf_data->second, similarity_cutoff);
	vector<article_data> article_data_processed;
	get_words_with_count_parallel(article_data_text, article_data_processed);
	g.add_texts(article_data_processed);
//	for(auto& t : text_to_chunks)
//		g.add_text(t);

//	for(int i = 0; i < texts.size(); ++i) {
//		cout << i << "\t";
//		for(string w : texts[i])
//			cout << w << " ";
//
//		cout << "\n";
//	}

	vector<vector<int>> res;
	res.reserve(g.get_groups().size());
	for(auto& a : g.get_groups())
		res.push_back(a.texts);

	return res;
}


vector<pair<pair<int, int>, float>> calc_similarities(const vector<tuple<vector<string>, string, string, double>>& article_data_text, const string& idf_file, float similarity_cutoff) {
	auto idf_data = loaded_idfs.at(idf_file);
	grouper g(idf_data->first, idf_data->second, similarity_cutoff);
	vector<article_data> article_data_processed;
	get_words_with_count_parallel(article_data_text, article_data_processed);
	g.add_texts(article_data_processed);
	vector<pair<pair<int, int>, float>> res;//(g.get_sims().begin(), g.get_sims().end());
	return res;
}


unordered_map<hash_t, pair<int, float>> get_chunks(const vector<string>& words, const string& idf_file) {
	auto idf_data = loaded_idfs.at(idf_file);
	auto& idf = idf_data->first;
	float max_idf = idf_data->second;
	unordered_map<hash_t, int> chunks;
	get_words_with_count(words, chunks);
	unordered_map<hash_t, pair<int, float>> res;
	for(auto& kvp : chunks) {
		auto it = idf.find(kvp.first);
		float c_idf = it == idf.end() ? max_idf : it->second;
		if (c_idf >= 1.f)
			res.emplace(kvp.first, make_pair(kvp.second, c_idf));
	}

	return res;
}

File: tgnews/groups.py
#    Copyright 2020 sorrge
#
#    This file is part of tg_news_cluster.
#
#    tg_news_cluster is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    tg_news_cluster is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with tg_news_cluster.  If not, see <https://www.gnu.org/licenses/>.
from datetime import datetime, timezone, timedelta
from typing import List, Tuple
import glob
import os

import news
from language import FileInfo, detect_languages, read_file_info

import libs.cpp_stuff as cpp


counts_for_grouping_ru = "data/chunk_counts_ru.bin"
counts_for_grouping_en = "data/chunk_counts_en.bin"


def calc_groups(texts: List[Tuple[List[str], str, str, float]], counts_file, similarity_cutoff):
    cpp.load_idf(counts_file)
    groups = cpp.group_texts(texts, counts_file, similarity_cutoff)
    return groups


def calc_groups_within_categories(texts: List[Tuple[List[str], str, str, float]], counts_file, similarity_cutoff, cats):
    all_cats = frozenset(cats) - frozenset(["junk"])
    cat_idx = [[i for i, c in enumerate(cats) if c == cat] for cat in all_cats]
    groups = []
    for idx in cat_idx:
        cat_texts = [texts[i] for i in idx]
        cat_groups = calc_groups(cat_texts, counts_file, similarity_cutoff)
        groups.extend([[idx[i] for i in group] for group in cat_groups])

    groups.sort(key=lambda x: len(x), reverse=True)
    return groups


def parse_datetime(datetime_string):
    try:
        return datetime.fromisoformat(datetime_string).astimezone(timezone.utc)
    except ValueError:
        return (datetime.utcnow() - timedelta(days=365*100)).astimezone(timezone.utc)


def extract_text_for_grouping(fi: FileInfo) -> List[str]:
    return (news.simple_tokenize_with_numbers(fi.title.lower()) + [""]) * 2 + \
           news.simple_tokenize_with_numbers(fi.text.lower())[:500]


def extract_texts_for_grouping(file_info) -> List[List[str]]:
    return [extract_text_for_grouping(fi) for fi in file_info]


def extract_infos_for_grouping(file_info):
    texts = extract_texts_for_grouping(file_info)
    return [(text, fi.site, fi.file, parse_datetime(fi.time).timestamp()) for text, fi in zip(texts, file_info)]


def group_news_ru(file_info: List[FileInfo], similarity_cutoff):
    info_for_grouping = extract_infos_for_grouping(file_info)
    categories = news.classify_ru(file_info)
    groups = calc_groups_within_categories(info_for_grouping, counts_for_grouping_ru, similarity_cutoff, categories)
    res = []
    aggregated_idx = []
    for group in groups:
        group.sort(key=lambda idx: parse_datetime(file_info[idx].time), reverse=True)
        group_title = file_info[group[0]].title
        res.append((group_title, [file_info[idx].file for idx in group]))
        aggregated_idx.extend(group)

    aggregated_idx = frozenset(aggregated_idx)
    for i, fi in enumerate(file_info):
        if i not in aggregated_idx and categories[i] != "junk":
            res.append((fi.title, [fi.file]))

    return res


def group_news_en(file_info: List[FileInfo], similarity_cutoff):
    info_for_grouping = extract_infos_for_grouping(file_info)
    categories = news.classify_en(file_info)
    groups = calc_groups_within_categories(info_for_grouping, counts_for_grouping_en, similarity_cutoff, categories)
    res = []
    aggregated_idx = []
    for group in groups:
        group.sort(key=lambda idx: parse_datetime(file_info[idx].time), reverse=True)
        group_title = file_info[group[0]].title
        res.append((group_title, [file_info[idx].file for idx in group]))
        aggregated_idx.extend(group)

    aggregated_idx = frozenset(aggregated_idx)
    for i, fi in enumerate(file_info):
        if i not in aggregated_idx and categories[i] != "junk":
            res.append((fi.title, [fi.file]))

    return res


def group_news(file_info, similarity_cutoff):
    lang_data = detect_languages(file_info)
    article_groups = []
    for ld in lang_data:
        if ld["lang_code"] == "ru":
            ru_files = frozenset(ld["articles"])
            ru_file_info = [fi for fi in file_info if fi.file in ru_files]
            groups = group_news_ru(ru_file_info, similarity_cutoff)
            for group_title, files in groups:
                article_groups.append({"title": group_title, "articles": files})
        elif ld["lang_code"] == "en":
            en_files = frozenset(ld["articles"])
            en_file_info = [fi for fi in file_info if fi.file in en_files]
            groups = group_news_en(en_file_info, similarity_cutoff)
            for group_title, files in groups:
                article_groups.append({"title": group_title, "articles": files})

    return article_groups


def group_news_in_folder(folder, similarity_cutoff):
    files = glob.glob(os.path.join(folder, "*.html"))
    file_info = read_file_info(files)
    res = group_news(file_info, similarity_cutoff)
    for ld in res:
        ld["articles"] = [os.path.basename(file) for file in ld["articles"]]

    return res
